### PR TITLE
TRD trapsim clean up

### PR DESCRIPTION
**Pull request by bazinski**
- Clean up trapsimulator code.
- Fix units.
- re do the generation of digits index.
- remove lots of output.
- remove a lot of debug statements.

### CODE REVIEW

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h
Comment (tdietel) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h]: Shouldn't this be debug level?

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
Comment (tdietel) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: Is the fix for trigger records still needed?

**Review by tdietel — DISMISSED** [review]
Looks good, I only spotted two minor issues.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h]: yes, I have noticed a fair amout of text output is still there, fixing.

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: No

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h
Comment (martenole) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h]: This line might be a leftover from a rebase? If not the assignment operator should simply be `= default` I think

File: Detectors/TRD/base/include/TRDBase/PadCalibrations.h
Comment (martenole) [inline Detectors/TRD/base/include/TRDBase/PadCalibrations.h]: Also a rebase artifact?

File: Detectors/TRD/base/macros/OCDB2CCDB.C
Comment (martenole) [inline Detectors/TRD/base/macros/OCDB2CCDB.C]: Again rebase... Otherwise you create hard to spot issues again ;)

**Review by martenole — COMMENTED** [review]
I didn't go through everything in detail but spotted some rebase artifacts

File: Detectors/TRD/base/macros/OCDB2CCDB.C
Comment (bazinski) [reply Detectors/TRD/base/macros/OCDB2CCDB.C]: thats weird, I dont remember touching that file.
Let me rather go through the diff closely, but yes of course this and those above need undoing.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h]: made default

File: Detectors/TRD/base/include/TRDBase/PadCalibrations.h
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/PadCalibrations.h]: fixed, well file reverted.

File: Detectors/TRD/base/macros/OCDB2CCDB.C
Comment (bazinski) [reply Detectors/TRD/base/macros/OCDB2CCDB.C]: fixed

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: removed

**Comment by sawenzel** [comment]
@martenole : Could you please re-review? (since you had comments)

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
Comment (martenole) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: Why is it actually 1077? I thought we have 521 installed chambers -> 1042 links, but maybe I miss something

**Comment by bazinski** [comment]
@sawenzel was going to say dont merge, something still wrong with tracklets, although not sure why the 2 failures since I have not touched the branch since is was all happy. I will come back after monday when raw stuff is done.

**Review by martenole — COMMENTED** [review]
Hi Sean,
thanks for including the small changes I requested.
I tried to test the `o2-trd-trap-sim` workflow with the changes from this PR and I get a seg fault. It seems to be originating from the digit sorting in `TRDTrapSimulatorSpec.cxx:335`. More details see below.
Also I got this message from the digit reader saying there are 9 entries in the trigger records although I simulated 10 events.. Is there still some problem?
`[134720:TRDDIGITREADER]: [20:01:42][INFO] TRDDigitReader digits size=5887 triggerrecords size=9 mc labels size (in bytes) = 78488`
Cheers,
Ole


[134721:TRAP]: The lines below might hint at the cause of the crash.
[134721:TRAP]: You may get help by asking at the ROOT forum http://root.cern.ch/forum
[134721:TRAP]: Only if you are really convinced it is a bug in ROOT then please submit a
[134721:TRAP]: report at http://root.cern.ch/bugs Please post the ENTIRE stack trace
[134721:TRAP]: from above as an attachment in addition to anything else
[134721:TRAP]: that might help us fixing this issue.
[134721:TRAP]: ===========================================================
[134721:TRAP]: #7  0x00007f005f983af4 in o2::trd::Digit::getDetector (this=0x55dfaf539f40) at /home/oschmidt/alice/sw/SOURCES/O2/v1.3.0/0/Detectors/TRD/base/include/TRDBase/Digit.h:52
[134721:TRAP]: #8  o2::trd::digitindexcompare (A=<optimized out>, B=1590043568, originalDigits=std::vector of length 5887, capacity 5887 = {...}) at /home/oschmidt/alice/sw/SOURCES/O2/v1.3.0/0/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:211
[134721:TRAP]: #9  0x00007f005f983ca4 in o2::trd::TRDDPLTrapSimulatorTask::<lambda(auto:26&&, auto:27&&)>::operator()<unsigned int&, unsigned int&> (__closure=<synthetic pointer>, PH1=
[134721:TRAP]: 0x55c507a30694: 32512, PH2=
[134721:TRAP]: 0x55c507a30690: 1590043568) at /home/oschmidt/alice/sw/SOURCES/O2/v1.3.0/0/Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:336
[134721:TRAP]: #10 __gnu_cxx::__ops::_Iter_comp_iter<o2::trd::TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext&)::<lambda(auto:26&&, auto:27&&)> >::operator()<__gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int> >, __gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int> > > (__it2=1590043568, __it1=32512, this=<synthetic pointer>) at /usr/include/c++/9/bits/predefined_ops.h:143
[134721:TRAP]: #11 std::__insertion_sort<__gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int> >, __gnu_cxx::__ops::_Iter_comp_iter<o2::trd::TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext&)::<lambda(auto:26&&, auto:27&&)> > >(__gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int, std::allocator<unsigned int> > >, __gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int, std::allocator<unsigned int> > >, __gnu_cxx::__ops::_Iter_comp_iter<o2::trd::TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext&)::<lambda(auto:26&&, auto:27&&)> >) (__first=__first
[134721:TRAP]: entry=1590043568, __last=..., __last
[134721:TRAP]: entry=21957, __comp=__comp
[134721:TRAP]: entry=...) at /usr/include/c++/9/bits/stl_algo.h:1846

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: oops should be 1080, more good reasons to "get 1077 from somewhere else" ;-) 
I left it as in run2 where all arrays/vectors are sized for 0-539 and 0-1079.
The hole is not at the end(is it) so its simpler to index all the way to 1080 and 540.

**Comment by bazinski** [comment]
@martenole with reference to the trigger records, when I looked it was fixed. @tdietel asked why the fix for the triggerrecords was still in the code, so I took it out.
Let me go check though, else will just put the fix back in.
As for the crash, the sorting code did not change, from when you and david ran it last time, so let me go look, I can only assume its the generation of the index.

**Comment by bazinski** [comment]
@martenole Are you running on an old simulation? An old simulation would have the corrupted triggrerecord.
if I look at that example David did with 10 PbPb, I see 10 triggerrecords. 
I also just did a simulation via 
$O2_ROOT/prodtests/sim_challenge.sh a trivial 10 event simulation and I get the following 

`[13191:TRDDIGITREADER]: [21:32:17][INFO] TRDDigitReader digits size=1958 triggerrecords size=4 mc labels size (in bytes) = 26396`
Something odd, maybe 6 piled up events?

**Comment by bazinski** [comment]
Crash is fixed, index to the digits is zero, now sized at instantiation of the vector.
There is also now a check to insure that the digitsindex is the same size as the digits themselves.

There is a problem retrieving save trapconfigs, patched for now.

**Comment by martenole** [comment]
Hi @bazinski 
thanks for fixing the crash. I see there are the right number of trigger records in the tracklets file now. Also the log message of the digits reader reports 10 trigger records.
Actually running now with the current dev branch (without this PR) I also sometimes see less trigger records than events simulated.. Is this already an effect of the busy time? I thought this was implemented so short at the moment that it should not make an effect. @jolopezl do you know why this is the case?
Cheers,
Ole

**Comment by jolopezl** [comment]
> Hi @bazinski
> thanks for fixing the crash. I see there are the right number of trigger records in the tracklets file now. Also the log message of the digits reader reports 10 trigger records.
> Actually running now with the current dev branch (without this PR) I also sometimes see less trigger records than events simulated.. Is this already an effect of the busy time? I thought this was implemented so short at the moment that it should not make an effect. @jolopezl do you know why this is the case?
> Cheers,
> Ole

Hi Ole, the current setup allows for pileup, and the times are not short enough to avoid this, so yes, you will see fewer trigger records than events simulated.

File: DataFormats/Detectors/TRD/src/RawData.cxx
Comment (tdietel) [inline DataFormats/Detectors/TRD/src/RawData.cxx]: I'm a bit rusty with operator precedence. Is this line the same as
`q0 | ( (q1 & 0xff) << 8 )`
I would find that more readable.

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
Comment (tdietel) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: Leftover debug statement?

File: Detectors/TRD/simulation/src/TrapSimulator.cxx
Comment (tdietel) [inline Detectors/TRD/simulation/src/TrapSimulator.cxx:320]: Wow, I wasn't aware this is in the TRAPsimulator. This looks like it replicates the output that one could get from a chamber running with a FeeServer. Could be useful, if we invest the time to set it up, e.g. at UCT.
However, it is somewhat complementatry to the tests that Venelin can do with a single readout board, and it is also not needed if we have optical readout of the chamber, e.g. at CERN.

For now, this is just a comment. If it is too much work to keep it up to date, it could be dropped.

**Review by tdietel — DISMISSED** [review]
I saw a few minor things, but from my side it's ready to be merged.

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: that sneaky little bugger slipped through. It would only appear once, now removed.

File: DataFormats/Detectors/TRD/src/RawData.cxx
Comment (bazinski) [reply DataFormats/Detectors/TRD/src/RawData.cxx]: The same yes, will bracket it.

File: Detectors/TRD/simulation/src/TrapSimulator.cxx
Comment (bazinski) [reply Detectors/TRD/simulation/src/TrapSimulator.cxx:320]: It should be correct. The only possible error is the assumption that the 0th cpu contains the header. I suspect this might not be reflective of the MCM. That assumption is made throughout though. It should not affect the output.

**Comment by davidrohr** [comment]
macos error unrelated, tested the slc7 O2/o2 CI locally and it is successful (and was green here before), merging.
@bazinski

**Comment by bazinski** [comment]
@davidrohr thanks.

**Pull request merged**
